internal/database/postgres/postgres.go

```diff
@@ -0,0 +1,265 @@
+		err = rows.Close()
+		if err != nil {
+			return errors.Join(err, tx.Rollback())
+		}
+
+		now := time.Now()
+		if lastSeen.Before(now) {
+			slog.Debug("last seen was before now, update it")
+			_, err = tx.Exec(`UPDATE Machines
+				SET LastSeen=$1
+				WHERE Hostname=$2`,
+				now, host)
+
+			if err != nil {
+				return errors.Join(err, tx.Rollback())
+			}
+		}
+	} else {
+		slog.Debug("row for this hostname doesn't exist, make it",
+			"hostname", host)
+
+		// Next() failing might be because of an error
+		err = rows.Err()
+		if err != nil {
+			return errors.Join(err, tx.Rollback())
+		}
+
+		// a machine with this hostname wasn't found, so make a new row
+		// for it
+		_, err = tx.Exec(`INSERT INTO Machines (Hostname, LastSeen)
+			VALUES ($1, $2)`,
+			host, time.Now())
+
+		if err != nil {
+			return errors.Join(err, tx.Rollback())
+		}
+
+		// TODO: in future, it should be added to a list to wait for
+		// approval
+	}
+
+	return tx.Commit()
+}
+
+func (conn postgresConn) AppendDataPoint(host string, packet uplink.GPUStats) error {
+	var err error
+
+	tx, err := conn.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	slog.Info("Find matching gpu")
+
+	// TODO: replace with Query
+	// This silently discards all rows other than the first
+	row := tx.QueryRow(`SELECT Id
+		FROM GPUs
+		WHERE Machine=$1`,
+		host)
+
+	var id int64
+	err = row.Scan(&id)
+
+	// check error type. If no rows found, add a new GPU
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			slog.Info("Didn't find a row, add new GPU")
+
+			// use TOFU approach, only set the name, brand, etc. on
+			// the first packet; future AppendDataPoint calls only
+			// add the 'dynamic' data to Stats
+			// TODO: reevaluate whether this is a good idea (what
+			// happens when a driver update occurs?)
+			newId := tx.QueryRow(`INSERT INTO GPUs
+				(Machine, Name, Brand, DriverVersion, MemoryTotal)
+				VALUES ($1, $2, $3, $4, $5)
+				RETURNING Id`,
+				host, packet.Name, packet.Brand,
+				packet.DriverVersion, packet.MemoryTotal)
+
+			err = newId.Scan(&id)
+			if err != nil {
+				return errors.Join(err, tx.Rollback())
+			}
+		} else {
+			slog.Info("Some other error occurred")
+			return errors.Join(err, tx.Rollback())
+		}
+	}
+
+	slog.Info("GPU now exists, push new stats")
+	now := time.Now()
+	_, err = tx.Exec(`INSERT INTO Stats
+		(Gpu, Recieved, MemoryUtilisation, GpuUtilisation, MemoryUsed,
+			FanSpeed, Temp)
+		VALUES ($1, $2, $3, $4, $5, $6, $7)`,
+		id, now, packet.MemoryUtilisation, packet.GPUUtilisation,
+		packet.MemoryUsed, packet.FanSpeed, packet.Temp)
+	if err != nil {
+		return errors.Join(err, tx.Rollback())
+	}
+
+	return tx.Commit()
+}
+
+func (conn postgresConn) LatestData() (map[string][]uplink.GPUStats, error) {
+	rows, err := conn.db.Query(`SELECT g.Machine, g.Name, g.Brand,
+			g.DriverVersion, g.MemoryTotal, s.MemoryUtilisation,
+			s.GpuUtilisation, s.MemoryUsed, s.FanSpeed, s.Temp
+		FROM GPUs g INNER JOIN Stats s ON g.Id = s.Gpu
```

Comment:
Trying to follow this query - are three inner joins necessary?

Reply:
I think so?

We want to get all the associated info for the latest entry in `Stats` (ie. the ones with the max recieved) so we:
- do a sub-query that aggregates every Stat row to get the largest recieved timestamp per stats.Gpu
- join this back with the Stats table to get the "dynamic" info for that row (ie. gpu usage, fan speed, etc.)
  - I don't believe we can do this in the aggregating sub-query because we don't want to aggregate gpu usage/fan speed, we just want the exact values from the row that's latest
 - then inner join that with the gpus table to get the "static" info (ie. driver version, machine hostname, etc.)

Reply:
I could add that explanation and move the sub-query so it's the first thing taken from to better illustrate that journey if that'd help?

Reply:
I might suggest this:

```sql
SELECT 
    g.Machine, 
    g.Name, 
    g.Brand,
    g.DriverVersion, 
    g.MemoryTotal, 
    s.MemoryUtilisation,
    s.GpuUtilisation, 
    s.MemoryUsed, 
    s.FanSpeed, 
    s.Temp
FROM GPUs g
INNER JOIN (
    SELECT *,
           ROW_NUMBER() OVER(PARTITION BY Gpu ORDER BY Recieved DESC) as rn
    FROM Stats
) s ON g.Id = s.Gpu
WHERE s.rn = 1
```